pinot-common/src/main/java/org/apache/pinot/common/lineage/SegmentLineage.java

```diff
@@ -70,6 +71,17 @@ public String addLineageEntry(LineageEntry lineageEntry) {
     return lineageId;
   }
 
+  /**
+   * Add lineage entry to the segment lineage metadata with the given lineage entry id
+   * @param lineageEntryId the id for the lineage entry
+   * @param lineageEntry a lineage entry
+   * @return the id for the input lineage entry for the access
+   */
+  public String addLineageEntry(String lineageEntryId, LineageEntry lineageEntry) {
```

Comment:
Keep only `String addLineageEntry(LineageEntry lineageEntry)` or `void addLineageEntry(String lineageEntryId, LineageEntry lineageEntry)`, don't keep both because it will be confusing. In other world, always generate UUID inside or outside of this class.

My suggestion is keeping `String addLineageEntry(LineageEntry lineageEntry)`, you can check existence of the `lineageId` (or not because UUID collision change is too low)

In order to update the lineage entry, you can add a method `void updateLineageEntry(String lineageEntryId, LineageEntry lineageEntry)`

Reply:
One reason I added `void addLineageEntry(String lineageEntryId, LineageEntry lineageEntry)` was because there's no easy way to pass `lineageEntryId` outside of the retry policy block. (`lineageEntryId`  is generated within retry function and it needs to be passed back to the top caller)

Due to the above reason, I have removed `String addLineageEntry(LineageEntry lineageEntry)` and kept `void addLineageEntry(String lineageEntryId, LineageEntry lineageEntry)`.

I also added `void updateLineageEntry(String lineageEntryId, LineageEntry lineageEntry) `